src/PromiseInterface.php

```diff
@@ -32,10 +33,11 @@ interface PromiseInterface
      *      than once.
      *  3. `$onProgress` (deprecated) may be called multiple times.
      *
-     * @param callable|null $onFulfilled
+     * @template TReturn of mixed
+     * @param callable(T): TReturn $onFulfilled
      * @param callable|null $onRejected
      * @param callable|null $onProgress This argument is deprecated and should not be used anymore.
-     * @return PromiseInterface
+     * @return (TReturn is PromiseInterface ? TReturn : PromiseInterface<TReturn>)
```

Comment:
How should we handle cases such as this?

```
$in = resolve(mt_rand(0, 100));
$out = $in->then(function (int $n): float {
    if ($n === 0) {
        throw new \UnexpectedValueException();
    }
    return 1 / $n;
});
```

I see you did some work regarding type assertions in https://github.com/phpstan/phpstan/discussions/7371, perhaps this is something we should include as part of this PR? :+1:

Reply:
Could do that, but then just the basics, and not the more complicated cases :+1: .